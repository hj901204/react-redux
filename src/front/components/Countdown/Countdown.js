import React from 'react'
import PropTypes from 'prop-types'
import {Button} from '../../bootstrap'

const propTypes = {
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  count: PropTypes.number,
  text: PropTypes.string
}

const defaultProps = {
  onClick: () => null,
  count: 60,
  disabled: false,
  text: '获取短信验证码'
}

class Countdown extends React.Component {
  state = {
    num: this.props.count,
    disabled: false
  }

  componentWillUnmount() {
    this.timer && clearInterval(this.timer)
  }

  handleClick = (e) => {
    this.props.onClick()
    this.setState({
      disabled: true,
      num: this.state.num - 1
    })
    this.timer = setInterval(() => {
      let disabled = this.state.disabled
      let num = this.state.num - 1
      if (num < 0) {
        disabled = false
        num = this.props.count
        clearInterval(this.timer)
      }
      this.setState({num, disabled})
    }, 1000)
  }

  render() {
    const {count, text, disabled, ...props} = this.props

    return (
      <Button
        {...props}
        type="button"
        disabled={this.state.disabled || disabled}
        className={this.props.className}
        onClick={this.handleClick}
      >{this.state.num !== count ? this.state.num + '秒后重发' : text}</Button>
    )
  }
}

Countdown.propTypes = propTypes
Countdown.defaultProps = defaultProps

export default Countdown
