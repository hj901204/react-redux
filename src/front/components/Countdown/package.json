{
  "name": "Countdown",
  "version": "1.0.0",
  "main": "./Countdown.js"
}
