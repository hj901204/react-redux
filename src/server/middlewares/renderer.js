const fse = require('fs-extra')
const path = require('path')

module.exports = () => async (ctx, next) => {
  const indexPath = path.resolve(__dirname, `../../${ctx.appId}/index.html`)
  let content = await fse.readFile(indexPath, 'utf8')

  if (typeof ctx.localeData === 'object') {
    const localeScripts = Object.keys(ctx.localeData).map(key => {
      return `window.${key}=${JSON.stringify(ctx.localeData[key])}`
    })
    content = content.replace(/(<!--state-->)/, `<script>${localeScripts.join(';')}</script>$1`)
  }

  ctx.body = content
}
