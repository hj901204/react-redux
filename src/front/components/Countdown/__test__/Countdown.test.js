import React from 'react'
import renderer from 'react-test-renderer'
import {render, mount} from 'enzyme'

import Countdown from '../Countdown'

describe('<Countdown />', () => {
  test('renders', () => {
    const tree = renderer.create(
      <Countdown />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('renders disabled', () => {
    const tree = renderer.create(
      <Countdown disabled />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('renders text', () => {
    const tree = renderer.create(
      <Countdown text="验证码" />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('click event', () => {
    const props = {
      count: 3,
      text: '获取短信验证码',
      onClick: jest.fn()
    }
    const countdown = mount(
      <Countdown {...props} />
    )

    expect(countdown.text()).toEqual(props.text)
    countdown.find('button').simulate('click')
    expect(props.onClick).toBeCalled()
    expect(countdown.text()).toEqual(`${props.count - 1}秒后重发`)
  })

  test('disabled click event', () => {
    const props = {
      count: 3,
      text: '获取短信验证码',
      disabled: true,
      onClick: jest.fn()
    }
    const countdown = mount(
      <Countdown {...props} />
    )

    expect(countdown.text()).toEqual(props.text)
    countdown.find('button').simulate('click')
    expect(props.onClick).not.toBeCalled()
    expect(countdown.text()).toEqual(props.text)
  })
})
