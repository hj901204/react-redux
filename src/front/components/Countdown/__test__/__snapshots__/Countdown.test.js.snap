// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Countdown /> renders 1`] = `
<button
  className="btn btn-default"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  获取短信验证码
</button>
`;

exports[`<Countdown /> renders disabled 1`] = `
<button
  className="btn btn-default disabled"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  获取短信验证码
</button>
`;

exports[`<Countdown /> renders text 1`] = `
<button
  className="btn btn-default"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  验证码
</button>
`;
